### PR TITLE
changing email

### PR DESCRIPTION
**Pull request by Sanjay1611**
Signed-off-by: Sanjay Nathani <sanjay.nathani@mayadata.io>

This PR contains following changes -

- Enter the email into `unverified_email` if a user wants to verify email.

- Move the `unverified_email` to `email` when user clicks on `Verify My Email` button on his mail id

This will ensure that `email` field will always contain the verified email of the user.

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -378,29 +378,31 @@ func (s *Server) SendVerificationLink(c *gin.Context, email string) {
 }
 
 // VerifyEmail marks a user email as verified
-func (s *Server) VerifyEmail(c *gin.Context) {
+func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
 	jwtUser, exists := c.Get(types.JWTUserCredentialsKey)
 	if !exists {
 		s.redirectError(c, errors.ErrInvalidAccessToken)
+		c.Redirect(http.StatusUnauthorized, redirectURL)
```
Comment (avishnu) [inline pkg/server/server.go:385]: Can you explain this change as code comments?

File: pkg/server/server.go [hunk 2]
```diff
@@ -378,29 +378,31 @@ func (s *Server) SendVerificationLink(c *gin.Context, email string) {
...
-	updatedUserInfo, err := usermanager.UpdateUserDetails(s.userStore, jwtUserCredentials)
-	if err != nil {
-		s.redirect(c, err)
-	}
-
-	tgr := &jwtmanager.TokenGenerateRequest{
-		UserInfo:       updatedUserInfo,
-		AccessTokenExp: time.Minute * 10,
+	if jwtUserCredentials.UnverifiedEmail != nil {
+		*jwtUserCredentials.Email = jwtUserCredentials.GetUnverifiedEmail()
```
Comment (avishnu) [inline pkg/server/server.go:391]: Lets add a success log here that Email xxx@yyycom is verified successfully.

Comment (avishnu) [inline pkg/server/server.go:391]: Is there any difference between jwtUserCredentials.UnverifiedEmail and jwtUserCredentials.GetUnverifiedEmail() ?

File: pkg/server/server.go [hunk 3]
```diff
@@ -378,29 +378,31 @@ func (s *Server) SendVerificationLink(c *gin.Context, email string) {
...
+		*jwtUserCredentials.UnverifiedEmail = ""
+	} else {
+		c.JSON(http.StatusBadRequest, gin.H{
+			"error": "No Email found to be verified",
```
Comment (avishnu) [inline pkg/server/server.go:395]: email

File: versionedController/v1/email/email.go [hunk 4]
```diff
@@ -45,18 +45,20 @@ func (emailController *EmailController) Post(c *gin.Context) {
 //Get verifies the email by a link
 func (emailController *EmailController) Get(c *gin.Context) {
 	token := c.Query("access")
+	redirectURL := types.PortalURL + "/verified-email"
 
 	jwtUserCredentials, err := controller.Server.GetUserFromToken(token)
 	if err != nil {
 		log.Error("Error occurred while parsing jwt token error: " + err.Error())
-		c.JSON(http.StatusUnauthorized, gin.H{
+		c.JSON(http.StatusInternalServerError, gin.H{
```
Comment (avishnu) [inline versionedController/v1/email/email.go:53]: Why this change?

File: versionedController/v1/email/email.go [hunk 5]
```diff
@@ -45,18 +45,20 @@ func (emailController *EmailController) Post(c *gin.Context) {
...
 			"error": err.Error(),
 		})
+		c.Redirect(http.StatusFound, redirectURL)
```
Comment (avishnu) [inline versionedController/v1/email/email.go:56]: This is error case, right? Is the status correct?

File: pkg/server/server.go (hunk 2 above)
Comment (Sanjay1611) [reply pkg/server/server.go:391]: `UnverifiedEmail` is a pointer to string and `GetUnverifiedEmail` returns its values as string and `""` otherwise

File: versionedController/v1/email/email.go (hunk 4 above)
Comment (Sanjay1611) [reply versionedController/v1/email/email.go:53]: This change has been pushed by mistake. Will rectify it.

File: pkg/server/server.go (hunk 2 above)
Comment (Sanjay1611) [reply pkg/server/server.go:392]: Have added the log in the end before redirecting the user

Comment (avishnu) [reply pkg/server/server.go:392]: Then, *jwtUserCredentials.UnverifiedEmail should give the value?
Main question is do we need an accessor method?

File: pkg/server/server.go (hunk 1 above)
Comment (avishnu) [reply pkg/server/server.go:386]: Trying to understand - why not redirect to error screen? Or am I missing something?

Comment (avishnu) [reply pkg/server/server.go:386]: I think its fine.

File: pkg/server/server.go [hunk 6]
```diff
@@ -378,29 +378,36 @@ func (s *Server) SendVerificationLink(c *gin.Context, email string) {
 }
 
 // VerifyEmail marks a user email as verified
-func (s *Server) VerifyEmail(c *gin.Context) {
+func (s *Server) VerifyEmail(c *gin.Context, redirectURL string) {
 	jwtUser, exists := c.Get(types.JWTUserCredentialsKey)
 	if !exists {
```
Comment (avishnu) [inline pkg/server/server.go:383]: This case may not arise as you are setting the key in line no. 60 of email.go. Is the understanding correct?

File: pkg/server/server.go (hunk 2 above)
Comment (avishnu) [reply pkg/server/server.go:392]: Ok, great

Comment (Sanjay1611) [reply pkg/server/server.go:392]: The accessor method will just return the zero value if we try to access it through a nil pointer. It just ensures that we do not get a nil pointer exception in worst case. 
Although we are already checking it in this case as we have to respond to UI accordingly.

File: pkg/server/server.go (hunk 1 above)
Comment (Sanjay1611) [reply pkg/server/server.go:386]: Okay. My thought process was to redirect the user to the same link and let UI take care if any error is there as UI will better know whether to redirect him to an error page (since we have quite a few possibility of errors ) or to the login page(if in case the token is not there).

File: pkg/server/server.go (hunk 6 above)
Comment (Sanjay1611) [reply pkg/server/server.go:383]: Yes @avishnu . The understanding is totally correct. I have added this check in case we need to call this method from some other place too where we do not set this key.

**Review by avishnu — APPROVED** [review]
lgtm
